File: src/main/java/alfheim/client/render/entity/RenderEntityAlfheimPixie.java
package alfheim.client.render.entity;

import static org.lwjgl.opengl.GL11.*;

import alfheim.api.lib.LibResourceLocations;
import alfheim.common.entity.EntityAlfheimPixie;
import net.minecraft.client.renderer.OpenGlHelper;
import net.minecraft.client.renderer.entity.RenderLiving;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.util.ResourceLocation;
import vazkii.botania.client.model.ModelPixie;

public class RenderEntityAlfheimPixie extends RenderLiving {

	public RenderEntityAlfheimPixie() {
		super(new ModelPixie(), 0.25F);
		setRenderPassModel(new ModelPixie());
		shadowSize = 0;
	}

	@Override
	protected ResourceLocation getEntityTexture(Entity entity) {
		return LibResourceLocations.pixie;
	}

	@Override
	public void doRender(Entity entity, double x, double y, double z, float yaw, float pitch) {
		if (entity.isInvisible()) return;
		glPushMatrix();
		glTranslated(0, -0.25, 0);
		super.doRender(entity, x, y, z, yaw, pitch);
		glPopMatrix();
	}

	protected int setPixieBrightness(EntityAlfheimPixie pixie, int par2, float par3) {
		if (par2 != 0)
			return -1;
		else {
			bindTexture(getEntityTexture(pixie));
			glEnable(GL_BLEND);
			glDisable(GL_ALPHA_TEST);
			glBlendFunc(GL_ONE, GL_ONE);

			if (pixie.isInvisible())
				glDepthMask(false);
			else
				glDepthMask(true);

			char c0 = 61680;
			float j = c0 % 65536;
			float k = c0 / 65536;
			OpenGlHelper.setLightmapTextureCoords(OpenGlHelper.lightmapTexUnit, j, k);
			glColor4d(1, 1, 1, 1);
			glColor4d(1, 1, 1, 1);
			return 1;
		}
	}

	@Override
	protected int shouldRenderPass(EntityLivingBase entity, int par2, float par3) {
		return setPixieBrightness((EntityAlfheimPixie) entity, par2, par3);
	}
}

File: src/main/java/alfheim/client/render/entity/RenderEntityAlfheimPixie.kt
package alfheim.client.render.entity

import alexsocol.asjlib.render.ASJRenderHelper
import alfheim.api.lib.LibResourceLocations
import alfheim.common.entity.EntityAlfheimPixie
import net.minecraft.client.renderer.entity.RenderLiving
import net.minecraft.entity.*
import org.lwjgl.opengl.GL11.*
import vazkii.botania.client.model.ModelPixie

object RenderEntityAlfheimPixie: RenderLiving(ModelPixie(), 0.25f) {
	
	init {
		setRenderPassModel(ModelPixie())
		shadowSize = 0f
	}
	
	override fun getEntityTexture(entity: Entity?) = LibResourceLocations.pixie
	
	override fun doRender(entity: Entity, x: Double, y: Double, z: Double, yaw: Float, pitch: Float) {
		if (entity.isInvisible) return
		
		glPushMatrix()
		glTranslatef(0f, -0.25f, 0f)
		super.doRender(entity, x, y, z, yaw, pitch)
		glPopMatrix()
		
		ASJRenderHelper.discard()
	}
	
	private fun setPixieBrightness(pixie: EntityAlfheimPixie, pass: Int): Int {
		if (pass != 0) return -1
		
		bindTexture(getEntityTexture(pixie))
		glEnable(GL_BLEND)
		glDisable(GL_ALPHA_TEST)
		glBlendFunc(GL_ONE, GL_ONE)
		
		if (pixie.isInvisible)
			glDepthMask(false)
		else
			glDepthMask(true)
		
		ASJRenderHelper.setGlow()
		glColor4f(1f, 1f, 1f, 1f)
		return 1
	}
	
	override fun shouldRenderPass(entity: EntityLivingBase, pass: Int, ticks: Float) = setPixieBrightness(entity as EntityAlfheimPixie, pass)
}
